dhall/src/Dhall/Map.hs

```diff
@@ -223,10 +221,10 @@ fromList [("C",3),("A",3)]
 insertWith :: Ord k => (v -> v -> v) -> k -> v -> Map k v -> Map k v
 insertWith f k v (Map m ks) = Map m' ks'
   where
-    m' = Data.Map.insertWith f k v m
+    (mayOldV, m') = Data.Map.insertLookupWithKey (\_k new old -> f new old) k v m
 
-    ks' | elem k ks = ks
-        | otherwise = k : ks
+    ks' | Just _ <- mayOldV = ks
+        | otherwise         = k : ks
```

Comment:
I feel like this only obfuscates what's going on. `insertLookupWithKey` is rather esoteric as far as I am concerned.

Reply:
>  `insertLookupWithKey` is rather esoteric as far as I am concerned.

It sure is!

The point is to avoid fiddling around with the keys list as much as possible. In practice we hardly ever (never?) insert keys that are already present in the map. This change simply increases laziness in the list.